src/main/BlockPool.cpp

```diff
@@ -0,0 +1,1216 @@
+    return data & ~uintptr_t{3};
+  }
+};
+
+struct Superblock;
+
+/**
+ * Freelists are used to store superblock nodes and large allocations.
+ * For large allocations the block and superblock members are unused.
+ */
+struct FreeListNode {
+  FreeListNode* next;
+  u32 block;  // Used only for superblock free nodes.
+  Superblock* superblock;  // Used only for superblock free nodes.
+};
+
+/**
+ * Superblocks are used to allocate small and medium sized objects.
+ * The bitset tracks which blocks are currently allocated.
+ * For small objects block_size is the actual object size.
+ * For medium objects block_size is the 2-log of the object size.
+ */
+struct Superblock {
+  u32 block_size;
+  u32 next_untouched;
+  FreeListNode* free_list;
+  u64 free[];  // Free bitset.
+
+  Superblock(u32 block_size, unsigned bitset_entries):
+      block_size(block_size),
+      next_untouched(0),
+      free_list(nullptr) {
+    // Here we mark all bitset entries as free, later we don't have to do
+    // precise range checking while iterating allocated blocks when
+    // the number of blocks is not evenly divisible by 64.
+    for (int i = 0; i < bitset_entries; ++i) {
+      free[i] = ~0ull;
+    }
+  }
+
+  /**
+   * Allocates a new superblock from the small object arena.
+   * Returns nullptr if the arena space is full.
+   * If nullptr is returned we need to fall back on using a separately
+   * allocated superblock.
+   */
+  static Superblock* NewSuperblock(int block_size);
+
+  /**
+   * Compute the address of the superblock header from a pointer.
+   * Returns nullptr if candidate pointer is not inside the small object arena
+   * or if the pointer points inside the superblock header.
+   */
+  static Superblock* SuperblockFromPointer(void* pointer);
+
+  /**
+   * Allocate a small block (bytes >= 32 && bytes <= 256).
+   * Returns nullptr if there is no more space left for this object
+   * size in the small object arena.
+   */
+  static void* AllocSmall(size_t block_size);
+
+  /**
+   * Free a pointer inside a given superblock. The block MUST be in the given
+   * superblock.
+   */
+  static void FreeSmallBlock(void* pointer, Superblock* superblock);
+
+  /** Allocate a medium or large object. */
+  static void* AllocLarge(unsigned size_log2);
+
+  /** Tag a block in a superblock as allocated. */
+  static void TagBlockAllocated(Superblock* superblock, unsigned block);
+};
+
+// Forward declarations for freelist functions (used in Superblock).
+static void add_free_block(uintptr_t data, unsigned size_log2);
+static void* remove_free_block(unsigned size_log2);
+
+// Pointers to small object superblocks with available blocks.
+static Superblock* small_superblocks[NUM_SMALL_POOLS];
+
+// Pointers to medium object superblocks with available blocks.
+static Superblock* medium_superblocks[18];
+
+/*
+ * Special-purpose hashtable implementation for tracking medium and large
+ * allocations.
+ *
+ * The table uses quadratic probing.
```

Comment:
The hashtable uses open addressing with quadratic probing to resolve collisions.